plugins/postgresql/reader.js

```diff
@@ -19,163 +19,180 @@ Reader.prototype.mostRecentWindow = function(from, to, next) {
-  SELECT ${what} from ${postgresUtil.table('candles')}
-  WHERE start <= ${to} AND start >= ${from}
-  ORDER BY start ASC
-  `));
-
-  var rows = [];
-  query.on('row', function(row) {
-    rows.push(row);
-  });
-
-  query.on('end',function(){
-    next(null, rows);
-  });
+    query.on('end',function(){
+      done();
+      next(null, rows);
+    });
+  });  
 }
 
 Reader.prototype.count = function(from, to, next) {
-  var query = this.db.query(new Query(`
-  SELECT COUNT(*) as count from ${postgresUtil.table('candles')}
-  WHERE start <= ${to} AND start >= ${from}
-  `));
-  var rows = [];
-  query.on('row', function(row) {
-    rows.push(row);
-  });
-
-  query.on('end',function(){
-    next(null, _.first(rows).count);
-  });
+  this.db.connect((err,client,done) => {
+    var query = client.query(new Query(`
+    SELECT COUNT(*) as count from ${postgresUtil.table('candles')}
+    WHERE start <= ${to} AND start >= ${from}
+    `));
+    var rows = [];
+    query.on('row', function(row) {
+      rows.push(row);
+    });
+
+    query.on('end',function(){
+      done();
+      next(null, _.first(rows).count);
+    });
+  });  
 }
 
 Reader.prototype.countTotal = function(next) {
-  var query = this.db.query(new Query(`
-  SELECT COUNT(*) as count from ${postgresUtil.table('candles')}
-  `));
-  var rows = [];
-  query.on('row', function(row) {
-    rows.push(row);
-  });
-
-  query.on('end',function(){
-    next(null, _.first(rows).count);
-  });
+  this.db.connect((err,client,done) => {
+    var query = client.query(new Query(`
+    SELECT COUNT(*) as count from ${postgresUtil.table('candles')}
+    `));
+    var rows = [];
+    query.on('row', function(row) {
+      rows.push(row);
+    });
+
+    query.on('end',function(){
+      done();
+      next(null, _.first(rows).count);
+    });
+  });  
 }
 
 Reader.prototype.getBoundry = function(next) {
-  var query = this.db.query(new Query(`
-  SELECT (
-    SELECT start
-    FROM ${postgresUtil.table('candles')}
-    ORDER BY start LIMIT 1
-  ) as first,
-  (
-    SELECT start
-    FROM ${postgresUtil.table('candles')}
-    ORDER BY start DESC
-    LIMIT 1
-  ) as last
-  `));
-  var rows = [];
-  query.on('row', function(row) {
-    rows.push(row);
-  });
-
-  query.on('end',function(){
-    next(null, _.first(rows));
-  });
+  this.db.connect((err,client,done) => {
+    var query = client.query(new Query(`
+    SELECT (
+      SELECT start
+      FROM ${postgresUtil.table('candles')}
+      ORDER BY start LIMIT 1
+    ) as first,
+    (
+      SELECT start
+      FROM ${postgresUtil.table('candles')}
+      ORDER BY start DESC
+      LIMIT 1
+    ) as last
+    `));
+    var rows = [];
+    query.on('row', function(row) {
+      rows.push(row);
+    });
+
+    query.on('end',function(){
+      done();
+      next(null, _.first(rows));
+    });
+  });  
 }
 
 Reader.prototype.close = function() {
-  this.db.end();
+  //obsolete due to connection pooling
+  //this.db.end();
```

Comment:
does this mean that the process can completely quiet without having to close anything? If yes please remove the completed Reader.prototype.close` function!

Reply:
Exactly, this is the goal of connection pooling. You do not close a connection anymore, instead you hand it back to the pool with callback done(). Not all DB plugins support this pooling method and therefore I think we can not remove this call from the core. For now, I left it in postgresql plugin code to do nothing, so it can be still called from core without any errors.

Reply:
:+1: